src/nvim/search.c

```diff
@@ -5106,3 +5106,527 @@ bool search_was_last_used(void)
+                                          nextMatch);
+  }
+
+  // Return best result
+  if (recursiveMatch && (!matched || bestRecursiveScore > *outScore)) {
+    // Recursive score is better than "this"
+    memcpy(matches, bestRecursiveMatches, maxMatches * sizeof(matches[0]));
+    *outScore = bestRecursiveScore;
+    return true;
+  } else if (matched) {
+    return true;  // "this" score is better than recursive
+  }
+
+  return false;  // no match
+}
+
+// fuzzy_match()
+//
+// Performs exhaustive search via recursion to find all possible matches and
+// match with highest score.
+// Scores values have no intrinsic meaning.  Possible score range is not
+// normalized and varies with pattern.
+// Recursion is limited internally (default=10) to prevent degenerate cases
+// (fuzpat="aaaaaa" str="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa").
+// Uses char_u for match indices. Therefore patterns are limited to MAXMATCHES
+// characters.
+//
+// Returns true if 'fuzpat' matches 'str'. Also returns the match score in
+// 'outScore' and the matching character positions in 'matches'.
+static int fuzzy_match(
+    char *str,
+    char *fuzpat,
+    int *outScore,
+    matchidx_T *matches,
+    int maxMatches)
+{
+  int recursionCount = 0;
+  int len = utf_char2len(*str);
+
+  *outScore = 0;
+
+  return fuzzy_match_recursive(
+      fuzpat, str, 0, outScore, str, len, NULL,
+      matches, maxMatches, 0, &recursionCount);
+}
+
+// Sort the fuzzy matches in the descending order of the match score.
+static int fuzzy_item_compare(const void *s1, const void *s2)
+{
+  int v1 = ((fuzzyItem_T *)s1)->score;
+  int v2 = ((fuzzyItem_T *)s2)->score;
+
+  return v1 == v2 ? 0 : v1 > v2 ? -1 : 1;
+}
+
+// Fuzzy search the string 'str' in a list of 'items' and return the matching
+// strings in 'fmatchlist'.
+// If 'items' is a list of strings, then search for 'str' in the list.
+// If 'items' is a list of dicts, then either use 'key' to lookup the string
+// for each item or use 'item_cb' Funcref function to get the string.
+// If 'retmatchpos' is true, then return a list of positions where 'str'
+// matches for each item.
+static void match_fuzzy(
+    list_T *items,
+    char *str,
+    char *key,
+    Callback *item_cb,
+    int retmatchpos,
+    list_T *fmatchlist)
+{
+  long len;
+  fuzzyItem_T *ptrs;
+  long i = 0;
+  bool found_match = false;
+  matchidx_T matches[MAXMATCHES];
+
+  len = tv_list_len(items);
+  if (len == 0) {
+    return;
+  }
+
+  ptrs = xcalloc(sizeof(fuzzyItem_T), len);
+  if (ptrs == NULL) {
+    return;
+  }
+
+  // For all the string items in items, get the fuzzy matching score
+  TV_LIST_ITER(items, li, {
+    int score;
+    char *itemstr;
+    typval_T rettv;
+    typval_T *tv_li = TV_LIST_ITEM_TV(li);
+
+    ptrs[i].item = li;
+    ptrs[i].score = SCORE_NONE;
+    itemstr = NULL;
+    rettv.v_type = VAR_UNKNOWN;
+
+    if (tv_check_str(tv_li)) {  // list of strings
+      itemstr = (char *)tv_li->vval.v_string;
```

Comment:
Here `tv_check_str()` checks that a value can be *cast* to a string, not that it *is* a string.

Reply:
Oh interesting